### PR TITLE
#231: statistics and algorithms overhaul

### LINKED ISSUE #231: Overhaul LBAF algorithms and statistics framework & update lbsStrictLocalizingCriterion
### TODO:
- `StrictLocalizingCriterion` is not updated as `TemperedCriterion` and does not work

Issue **TODO** expanded by @ppebay to include several other aspects such as:
* generalizing of the global statistics selection and mapping to respective computation methods
* formatting/documentation of algorithms that is not always consistent
* make abstract descriptive statistics (e.g. min, max, ave, etc.) by hiding statistics computation return index behind a human-legible key (e.g. `lbsStatistics.MIN` etc.)
* improve `LBAF` statistical and plain text reporting

### PR DESCRIPTION
**Pull request by ppebay**
Resolves #231

### CODE REVIEW

**Review by lifflander — APPROVED** [review]
This looks like a definite improvement to me!

File: src/lbaf/Execution/lbsAlgorithmBase.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import sys
 from logging import Logger
 
-from ..IO.lbsStatistics import compute_function_statistics
+from ..IO.lbsStatistics import compute_function_statistics, Statistics
```
Comment (cz4rs) [inline src/lbaf/Execution/lbsAlgorithmBase.py:5]: Remove unused import:
```suggestion
from ..IO.lbsStatistics import compute_function_statistics
```

File: src/lbaf/Execution/lbsRuntime.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 from ..Model.lbsWorkModelBase import WorkModelBase
 from ..Execution.lbsAlgorithmBase import AlgorithmBase
-from ..IO.lbsStatistics import compute_function_statistics, min_Hamming_distance
+from ..IO.lbsStatistics import Statistics, compute_function_statistics, min_Hamming_distance
```
Comment (cz4rs) [inline src/lbaf/Execution/lbsRuntime.py:6]:
```suggestion
from ..IO.lbsStatistics import compute_function_statistics, min_Hamming_distance
```

File: src/lbaf/IO/lbsStatistics.py [hunk 3]
```diff
@@ -1,12 +1,43 @@
-from logging import Logger
+import enum
```
Comment (cz4rs) [inline src/lbaf/IO/lbsStatistics.py:1]: Unused:
```suggestion
```

**Review by cz4rs — APPROVED** [review]
Looks good to me. Minor cleanup of imports suggested in the comments.

**Comment by cz4rs** [comment]
I am getting lots of `blank line contains whitespace` warnings when viewing the code locally. You might want to enable "Trim trailing whitespace on save" or whatever similar option your editor has :slightly_smiling_face: 

Also have you considered adding a basic unit test for the new `Statistics` class? (I'm thinking mostly derived statistics)

File: src/lbaf/IO/lbsStatistics.py [hunk 4]
```diff
@@ -1,12 +1,43 @@
...
 import math
 import random as rnd
-
+from logging import Logger
 from numpy import random
 
+class Statistics:
+    """A class storing descriptive statistics."""
+
+    def __init__(self, n, mini, mean, maxi, var, g1, g2):
```
Comment (cz4rs) [inline src/lbaf/IO/lbsStatistics.py:10]: Maybe type hints here as well?

**Pull request merged**